src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        if (typeof action.extraData === 'object' && action.extraData !== null) {
+          Object.assign(currentState, action.extraData);
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+      default:
+        break;
+    }
+    stateHistory.push({ ...currentState });
```

Comment:
Pushing the copy of the current state to the stateHistory at the end of each loop iteration is correctly placed outside of the switch block. This follows the DRY principle and ensures that the operation is performed only once per iteration.